File: python/src/resp_bench/client/factory.py
"""Driver registry: maps ``driver_id`` to a client implementation.

Implementations are imported lazily so that ``--info`` and unit tests do not
require heavy optional dependencies (valkey-glide, redis) to be installed when
a given driver is not used.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Callable, Dict

from ..config.driver_config import DriverConfig

if TYPE_CHECKING:  # pragma: no cover
    from .benchmark_client import AsyncBenchmarkClient


def _make_glide() -> "AsyncBenchmarkClient":
    from .impl.glide_client import GlideBenchmarkClient

    return GlideBenchmarkClient()


def _make_redis_py() -> "AsyncBenchmarkClient":
    from .impl.redis_py_client import RedisPyClient

    return RedisPyClient()


def _make_valkey_py() -> "AsyncBenchmarkClient":
    from .impl.valkey_py_client import ValkeyPyClient

    return ValkeyPyClient()


def _make_recording() -> "AsyncBenchmarkClient":
    from .impl.recording_client import RecordingClient

    return RecordingClient()


class BenchmarkClientFactory:
    # Ordered so --info lists them predictably.
    _FACTORIES: Dict[str, Callable[[], "AsyncBenchmarkClient"]] = {
        "valkey-glide-python": _make_glide,
        "redis-py": _make_redis_py,
        "valkey-py": _make_valkey_py,
        "recording": _make_recording,
    }

    @classmethod
    def supported_drivers(cls) -> list[str]:
        return list(cls._FACTORIES.keys())

    @classmethod
    def create(cls, driver_id: str) -> "AsyncBenchmarkClient":
        key = (driver_id or "").lower()
        factory = cls._FACTORIES.get(key)
        if factory is None:
            raise ValueError(
                f"Unknown driver: {driver_id}. "
                f"Supported: {', '.join(cls._FACTORIES)}"
            )
        return factory()

    @classmethod
    async def create_and_connect(
        cls, host: str, port: int, config: DriverConfig
    ) -> "AsyncBenchmarkClient":
        client = cls.create(config.driver_id)
        await client.connect(host, port, config)
        return client

File: php/src/Client/Factory.php
<?php

declare(strict_types=1);

namespace RespBench\Client;

use InvalidArgumentException;
use RespBench\Client\Impl\PhpRedisClient;
use RespBench\Client\Impl\RecordingClient;
use RespBench\Client\Impl\ValkeyGlidePhpClient;
use RespBench\Config\DriverConfig;

/**
 * Factory for creating benchmark client instances by driver id.
 */
final class Factory
{
    /** @var array<string,class-string<BenchmarkClient>> */
    private const DRIVERS = [
        'valkey-glide-php' => ValkeyGlidePhpClient::class,
        'phpredis' => PhpRedisClient::class,
        'recording' => RecordingClient::class,
    ];

    public static function create(string $driverId): BenchmarkClient
    {
        $class = self::DRIVERS[$driverId] ?? null;
        if ($class === null) {
            throw new InvalidArgumentException(
                "Unknown driver: {$driverId}. Supported: " . implode(', ', array_keys(self::DRIVERS))
            );
        }

        return new $class();
    }

    public static function createAndConnect(string $host, int $port, DriverConfig $config): BenchmarkClient
    {
        $client = self::create((string) $config->driverId);
        $client->connect($host, $port, $config);

        return $client;
    }

    /**
     * @return list<string>
     */
    public static function supportedDrivers(): array
    {
        return array_keys(self::DRIVERS);
    }
}
